src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,222 @@
     @Override
     public void add(T value, int index) {
+        ListNode newnode = new ListNode<>(value);
+        if (size == 0 && index == 0) {
+            head = newnode;
+            tail = newnode;
+            size++;
+            return;
+        }
+        if (index == 0 && size > 0) {
+            newnode.setNext(head);
+            head.setPrev(newnode);
+            head = newnode;
+            size++;
+            return;
+        }
+        if (size == index) {
+            tail.setNext(newnode);
+            newnode.setPrev(tail);
+            tail = newnode;
+            size++;
+            return;
+        }
+
+        ListNode temp = findElementByIndex(index);
+        newnode.setNext(temp);
+        newnode.setPrev(temp.getPrev());
+        temp.setPrev(newnode);
+        newnode.getPrev().setNext(newnode);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        ListNode temp = findElementByIndex(index);
+        return (T) temp.getValue();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        ListNode temp = findElementByIndex(index);
+        T oldvalue = (T) temp.getValue();
+        temp.setValue(value);
+        return oldvalue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        ListNode temp = findElementByIndex(index);
+        unlink(temp);
+        return (T) temp.getValue();
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        //throw exception if empty ?
+        if (size == 0) {
+            return false;
+        }
+        boolean found = false;
+        ListNode temp = head;
+        if (temp.getValue() != null && temp.getValue().equals(object)
+                || temp.getValue() == object) {
+            unlink(head);
+            found = true;
+            return found;
+        }
+
+        while (temp != tail && !found) {
+            if (temp.getValue() != null && temp.getValue().equals(object)
+                    || temp.getValue() == object) {
+                unlink(temp);
+                found = true;
+            }
+            if (found) {
+                return found;
+            }
+            temp = temp.getNext();
+        }
+        if (temp.getValue() != null && temp.getValue().equals(object)
+                || temp.getValue() == object) {
+            unlink(tail);
+            found = true;
+
+        }
+        return found;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private ListNode findElementByIndex(int index) {
+        if (index < 0 || index > size - 1 || isEmpty()) {
+            throw new IndexOutOfBoundsException();
```

Comment:
```suggestion
            throw new IndexOutOfBoundsException();
```
we always write messages when throw exceptions